“cannot initialize a tuple struct which contains private fields” with no tuple struct involved

### Code

```Rust
struct Foo {}
struct Bar {}

impl Foo {
    fn into_bar(self) -> Bar {
        Bar(self)
    }
}
```

### Current output

```Shell
error[E0423]: cannot initialize a tuple struct which contains private fields
 --> src/lib.rs:6:9
  |
6 |         Bar(self)
  |         ^^^
  |
  = note: constructor is not visible here due to private fields
```

### Desired output

```Shell
error[E0423]: expected function, tuple struct or tuple variant, found struct `Bar`
 --> src/lib.rs:6:9
  |
6 |         Bar(self)
  |         ^^^
```

### Rationale and extra context

The message is incorrect in two ways:

* `Bar` is not a tuple struct
* `Bar` has no private fields

It appears that something is making too many assumptions about why a tuple struct constructor might not be visible at a specific call site.

### Other cases

The incorrect diagnostic only occurs for `Bar(self)`, and not when a normal variable is used in place of `self`. It also seems like sometimes the correct error is printed, nondeterministically, but I have not found consistent reproduction conditions.

### Rust Version

```Shell
rustc 1.93.0-nightly (01867557c 2025-11-12)
binary: rustc
commit-hash: 01867557cd7dbe256a031a7b8e28d05daecd75ab
commit-date: 2025-11-12
host: aarch64-apple-darwin
release: 1.93.0-nightly
LLVM version: 21.1.5
```

### Anything else?

@rustbot label +regression-from-stable-to-nightly

--- a/compiler/rustc_resolve/src/late/diagnostics.rs
+++ b/compiler/rustc_resolve/src/late/diagnostics.rs
@@ -823,8 +823,14 @@ impl<'ast, 'ra, 'tcx> LateResolutionVisitor<'_, 'ast, 'ra, 'tcx> {
                 }
 
                 if let Some(Res::Def(DefKind::Struct, def_id)) = res {
-                    if self.has_private_fields(def_id) {
+                    let private_fields = self.has_private_fields(def_id);
+                    let adjust_error_message =
+                        private_fields && self.is_struct_with_fn_ctor(def_id);
+                    if adjust_error_message {
                         self.update_err_for_private_tuple_struct_fields(err, &source, def_id);
+                    }
+
+                    if private_fields {
                         err.note("constructor is not visible here due to private fields");
                     }
                 } else {
@@ -1644,15 +1650,17 @@ impl<'ast, 'ra, 'tcx> LateResolutionVisitor<'_, 'ast, 'ra, 'tcx> {
         }
     }
 
-    fn has_tuple_fields(&mut self, def_id: DefId) -> bool {
-        // As we cannot name a field "0", this is an indictation
-        // that we are looking at a tuple struct
-        if let Some(fields) = self.r.field_idents(def_id) {
-            if let Some(first_field) = fields.get(0) {
-                return first_field.name.as_str() == "0";
-            }
-        }
-        false
+    fn is_struct_with_fn_ctor(&mut self, def_id: DefId) -> bool {
+        def_id
+            .as_local()
+            .and_then(|local_id| self.r.struct_constructors.get(&local_id))
+            .map(|struct_ctor| {
+                matches!(
+                    struct_ctor.0,
+                    def::Res::Def(DefKind::Ctor(CtorOf::Struct, CtorKind::Fn), _)
+                )
+            })
+            .unwrap_or(false)
     }
 
     fn update_err_for_private_tuple_struct_fields(
@@ -1677,18 +1685,17 @@ impl<'ast, 'ra, 'tcx> LateResolutionVisitor<'_, 'ast, 'ra, 'tcx> {
                 span: call_span,
                 ..
             })) => {
-                if self.has_tuple_fields(def_id) {
-                    err.primary_message(
-                        "cannot initialize a tuple struct which contains private fields",
-                    );
-                    self.suggest_alternative_construction_methods(
-                        def_id,
-                        err,
-                        path.span,
-                        *call_span,
-                        &args[..],
-                    );
-                }
+                err.primary_message(
+                    "cannot initialize a tuple struct which contains private fields",
+                );
+                self.suggest_alternative_construction_methods(
+                    def_id,
+                    err,
+                    path.span,
+                    *call_span,
+                    &args[..],
+                );
+
                 self.r
                     .field_idents(def_id)
                     .map(|fields| fields.iter().map(|f| f.span).collect::<Vec<_>>())

--- a/compiler/rustc_resolve/src/late/diagnostics.rs
+++ b/compiler/rustc_resolve/src/late/diagnostics.rs
@@ -823,8 +823,10 @@ impl<'ast, 'ra, 'tcx> LateResolutionVisitor<'_, 'ast, 'ra, 'tcx> {
                 }
 
                 if let Some(Res::Def(DefKind::Struct, def_id)) = res {
-                    self.update_err_for_private_tuple_struct_fields(err, &source, def_id);
-                    err.note("constructor is not visible here due to private fields");
+                    if self.has_private_fields(def_id) {
+                        self.update_err_for_private_tuple_struct_fields(err, &source, def_id);
+                        err.note("constructor is not visible here due to private fields");
+                    }
                 } else {
                     err.span_suggestion(
                         call_span,
@@ -1642,6 +1644,17 @@ impl<'ast, 'ra, 'tcx> LateResolutionVisitor<'_, 'ast, 'ra, 'tcx> {
         }
     }
 
+    fn has_tuple_fields(&mut self, def_id: DefId) -> bool {
+        // As we cannot name a field "0", this is an indictation
+        // that we are looking at a tuple struct
+        if let Some(fields) = self.r.field_idents(def_id) {
+            if let Some(first_field) = fields.get(0) {
+                return first_field.name.as_str() == "0";
+            }
+        }
+        false
+    }
+
     fn update_err_for_private_tuple_struct_fields(
         &mut self,
         err: &mut Diag<'_>,
@@ -1664,17 +1677,18 @@ impl<'ast, 'ra, 'tcx> LateResolutionVisitor<'_, 'ast, 'ra, 'tcx> {
                 span: call_span,
                 ..
             })) => {
-                err.primary_message(
-                    "cannot initialize a tuple struct which contains private fields",
-                );
-                self.suggest_alternative_construction_methods(
-                    def_id,
-                    err,
-                    path.span,
-                    *call_span,
-                    &args[..],
-                );
-                // Use spans of the tuple struct definition.
+                if self.has_tuple_fields(def_id) {
+                    err.primary_message(
+                        "cannot initialize a tuple struct which contains private fields",
+                    );
+                    self.suggest_alternative_construction_methods(
+                        def_id,
+                        err,
+                        path.span,
+                        *call_span,
+                        &args[..],
+                    );
+                }
                 self.r
                     .field_idents(def_id)
                     .map(|fields| fields.iter().map(|f| f.span).collect::<Vec<_>>())

--- a/tests/ui/resolve/regression-struct-called-as-function-148919.rs
+++ b/tests/ui/resolve/regression-struct-called-as-function-148919.rs
@@ -1,0 +1,10 @@
+struct Bar {}
+
+impl Bar {
+    fn into_self(self) -> Bar {
+        Bar(self)
+        //~^ ERROR expected function, tuple struct or tuple variant, found struct `Bar` [E0423]
+    }
+}
+
+fn main() {}

--- a/tests/ui/resolve/regression-struct-called-as-function-148919.stderr
+++ b/tests/ui/resolve/regression-struct-called-as-function-148919.stderr
@@ -1,0 +1,9 @@
+error[E0423]: expected function, tuple struct or tuple variant, found struct `Bar`
+  --> $DIR/regression-struct-called-as-function-148919.rs:5:9
+   |
+LL |         Bar(self)
+   |         ^^^
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0423`.

